contrib/babelfishpg_tsql/src/procedures.c

```diff
@@ -1107,171 +1114,174 @@ sp_describe_undeclared_parameters_internal(PG_FUNCTION_ARGS)
-			int			colname_len;
-
-			lc1 = list_nth_cell(target_attnums, target_attnum_i);
-			undeclaredparams->targetattnums[num_target_attnums] = lfirst_int(lc1);
-
-			col = (ResTarget *) list_nth(cols, target_attnum_i);
-			colname_len = strlen(col->name);
-			undeclaredparams->targetcolnames[num_target_attnums] = (char *) palloc(NAMEDATALEN);
-			strncpy(undeclaredparams->targetcolnames[num_target_attnums], col->name, NAMEDATALEN);
-			undeclaredparams->targetcolnames[num_target_attnums][colname_len] = '\0';
+			undeclaredparams->tablename = (char *) palloc(NAMEDATALEN);
+			relname_len = strlen(relation->relname);
+			strncpy(undeclaredparams->tablename, relation->relname, NAMEDATALEN);
+			undeclaredparams->tablename[relname_len] = '\0';
+			undeclaredparams->schemaoid = RelationGetNamespace(r);
+			undeclaredparams->reloid = RelationGetRelid(r);
+
+			undeclaredparams->targetattnums = (int *) palloc(sizeof(int) * list_length(target_attnums));
+			undeclaredparams->targetcolnames = (char **) palloc(sizeof(char *) * list_length(target_attnums));
+		
+
+			/* Record attnums and column names of the target table */
+			target_attnum_i = 0;
+			target_attnums_len = list_length(target_attnums);
+			while (target_attnum_i < target_attnums_len)
+			{
+				ListCell   *lc1;
+				ResTarget  *col;
+				int			colname_len;
 
-			target_attnum_i += 1;
-			num_target_attnums += 1;
-		}
+				lc1 = list_nth_cell(target_attnums, target_attnum_i);
+				undeclaredparams->targetattnums[num_target_attnums] = lfirst_int(lc1);
 
-		relation_close(r, AccessShareLock);
-		pfree(pstate);
+				col = (ResTarget *) list_nth(cols, target_attnum_i);
+				colname_len = strlen(col->name);
+				undeclaredparams->targetcolnames[num_target_attnums] = (char *) palloc(NAMEDATALEN);
+				strncpy(undeclaredparams->targetcolnames[num_target_attnums], col->name, NAMEDATALEN);
+				undeclaredparams->targetcolnames[num_target_attnums][colname_len] = '\0';
 
-		/*
-		 * Parse the list of parameters, and determine which and how many are
-		 * undeclared.
-		 */
-		switch (nodeTag(parsetree->stmt))
-		{
-			case T_InsertStmt:
-				select_stmt = (SelectStmt *) insert_stmt->selectStmt;
-				values_list = select_stmt->valuesLists;
-				break;
-			case T_UpdateStmt:
+				target_attnum_i += 1;
+				num_target_attnums += 1;
+			}
 
-				/*
-				 * In an UPDATE statement, we could have both SET and WHERE
-				 * with undeclared parameters. That's targetList (SET ...) and
-				 * whereClause (WHERE ...)
-				 */
-				values_list = list_make1(handle_where_clause_restargets_right(pstate, update_stmt->whereClause, update_stmt->targetList));
-				break;
-			case T_DeleteStmt:
-				values_list = list_make1(handle_where_clause_restargets_right(pstate, delete_stmt->whereClause, NIL));
-				break;
-			default:
-				ereport(ERROR,
-						(errcode(ERRCODE_FEATURE_NOT_SUPPORTED),
-						 errmsg("Unsupported use case in sp_describe_undeclared_parameters")));
-				break;
-		}
+			relation_close(r, AccessShareLock);
```

Comment:
Do we have guarantee here that relation was opened? What if is_supported_case_sp_describe_undeclared_parameters = false (ie nodeTag is default).